compiler/src/lnntoamm/Types.ts

```diff
@@ -278,84 +550,127 @@ export class FunctionType extends Type {
     const matrix: Array<Type[]> = args.map((arg) => {
       return arg.fnselectOptions();
     });
+    isDbg && console.log('matrix:', matrix);
     // TODO: this weight system feels like it can be inaccurate
     // the weight of a particular function is computed by the sum
     // of the indices in each dimension, with the highest sum
     // having the greatest preference
-    const fnsByWeight = new Map<number, [Fn, Type][]>();
-    const indices = matrix.map(() => 0);
+    const fnsByWeight = new Map<number, [Fn, Type[], Type][]>();
+    const getIndices = matrixIndices(matrix);
     // keep it as for instead of while for debugging reasons
-    for (let i = 0; ; i++) {
+    for (
+      let indicesRes = getIndices.next();
+      !indicesRes.done;
+      indicesRes = getIndices.next()
+    ) {
```

Comment:
Same comment here, this looks iterable, and I think this would work:

```suggestion
    for (let indicesRes of getIndices) {
```